From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Frediano Ziglio <fziglio@redhat.com>
Date: Sat, 21 Jan 2017 09:01:18 +0000
Subject: [spice-server] stream-channel: Do not show an empty blank screen on
 start

Start showing something when we have a surface and stream
instead of showing a blank screen which is now not useful.
Was useful for debugging purposes to understand that the
new channel was sending messages correctly to client and
client could handle them.

Signed-off-by: Frediano Ziglio <fziglio@redhat.com>
Acked-by: Jonathon Jongsma <jjongsma@redhat.com>
---
 server/stream-channel.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/server/stream-channel.c b/server/stream-channel.c
index 2ad9ebae3..e89563b82 100644
--- a/server/stream-channel.c
+++ b/server/stream-channel.c
@@ -360,8 +360,13 @@ stream_channel_connect(RedChannel *red_channel, RedClient *red_client, RedsStrea
     red_channel_client_ack_zero_messages_window(rcc);
 
     // "emulate" dcc_start
-    // TODO only if "surface"
     red_channel_client_pipe_add_empty_msg(rcc, SPICE_MSG_DISPLAY_INVAL_ALL_PALETTES);
+
+    // only if "surface"
+    if (channel->width == 0 || channel->height == 0) {
+        return;
+    }
+
     // pass proper data
     red_channel_client_pipe_add_type(rcc, RED_PIPE_ITEM_TYPE_SURFACE_CREATE);
     // surface data
@@ -407,8 +412,8 @@ static void
 stream_channel_init(StreamChannel *channel)
 {
     channel->stream_id = -1;
-    channel->width = 1024;
-    channel->height = 768;
+    channel->width = 0;
+    channel->height = 0;
 }
 
 void
@@ -421,11 +426,14 @@ stream_channel_change_format(StreamChannel *channel, const StreamMsgFormat *fmt)
 
     // send new create surface if required
     if (channel->width != fmt->width || channel->height != fmt->height) {
+        if (channel->width != 0 && channel->height != 0) {
+            red_channel_pipes_add_type(red_channel, RED_PIPE_ITEM_TYPE_SURFACE_DESTROY);
+        }
         channel->width = fmt->width;
         channel->height = fmt->height;
-        red_channel_pipes_add_type(red_channel, RED_PIPE_ITEM_TYPE_SURFACE_DESTROY);
         red_channel_pipes_add_type(red_channel, RED_PIPE_ITEM_TYPE_SURFACE_CREATE);
         // TODO monitors config ??
+        red_channel_pipes_add_empty_msg(red_channel, SPICE_MSG_DISPLAY_MARK);
     }
 
     // allocate a new stream id
